Remove hardcoded node version

WhyNotHugo opened the pull request:
The install script installed node 11, but the application requires node
14, so build fails.

This changeset removes the hardcoded version from the build script, and
simply uses the version in .nvmrc.

WhyNotHugo commented:
This actually only prevents the build from crashing early on, but it still freezes near the end. I'm unsure why that is.

noroadsleft commented:
@WhyNotHugo Do you want to try to track down why the build is freezing, or would you like me to merge this as-is?

I can't test this because I don't/can't run Docker.

yanfali commented:
Wait, if it's not building then this shouldn't be merged. I don't have time
to look at it right now

On Tue, Nov 9, 2021, 13:25 James Young ***@***.***> wrote:

> @WhyNotHugo <https://github.com/WhyNotHugo> Do you want to try to track
> down why the build is freezing, or would you like me to merge this as-is?
>
> I can't test this because I don't/can't run Docker.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/qmk/qmk_configurator/pull/1015#issuecomment-964561666>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARLSUZ7SWGENABLJJLHW6LULGGUXANCNFSM5HPXUMNQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

WhyNotHugo commented:
`master` simply crashes mid-way, because it installs node11 but then tries to use node14.

This changeset fixes that crash. When I tried it the other day it seemed to freeze at the end, but I left it several minutes and it finished today (4m12s in total).

I can add a follow-up PR to build the docker image (and push it somewhere) if you want more confidence in the Dockerfile's stability.

noroadsleft commented:
@yanfali The non-building apparently only affects Docker; I can run it locally as normal.

yanfali commented:
Since this is only for docker it matters

On Tue, Nov 9, 2021 at 2:09 PM James Young ***@***.***> wrote:

> @yanfali <https://github.com/yanfali> The non-building apparently only
> affects Docker; I can run it locally as normal.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/qmk/qmk_configurator/pull/1015#issuecomment-964590985>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARLSU644M2IHPAO4BEXZVDULGLYXANCNFSM5HPXUMNQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
"We do not learn so much by our successes as we learn by failures -- our
own and others'.  Especially if we see the failures properly corrected."
-- Frank Lloyd Wright

WhyNotHugo commented:
You're very right; that script is only used for building the docker image. I've dropped in the script in favour of in-lining the code in it. I've also switched the image to simplify that build process (see commit description).

I've also added a CI job to prevent regressions.

yanfali commented:
I'll try to look this over this evening PST time.

yanfali commented on Dockerfile:
```diff
@@ -1,31 +1,18 @@
 # Build configurator
-RUN /bin/bash -i /qmk_configurator/bin/docker_build.sh
+RUN yarn install
+ENV VUE_APP_API_URL="/api yarn run build"
```
I know this is from the original but it looks weird. I wonder if this is one of this things that accidentally works and it should be VUE_APP_API_URL="/api" # yarn run build? because it doesn't really make sense in the way it's used to build URLs. @skullydazed was this intentional?

skullydazed replied:
The original did not have quotes, it was a command from a shell script that would set the environment var `VUE_APP_API_URL` to `/api` and then run the command `yarn run build`. I think that should be this:

```suggestion
RUN VUE_APP_API_URL=/api yarn run build
```

WhyNotHugo commented:
The included nginx config proxies `/api` to an existing API instance by default, so having Vue try to communicate with the API assuming it runs on the same domain makes sense; since nginx then does the proxying.

This also ignores CORS checks, since it's all on the same site.

On Wed, 10 Nov 2021, at 18:02, Yan-Fa Li wrote:
> 
> 
> ***@***.**** commented on this pull request.
> 
> 
> In Dockerfile <https://github.com/qmk/qmk_configurator/pull/1015#discussion_r746853307>:
> 
> >  
>  # Build configurator
> -RUN /bin/bash -i /qmk_configurator/bin/docker_build.sh
> +RUN yarn install
> +ENV VUE_APP_API_URL="/api yarn run build"
> 
> I know this is from the original but it looks weird. I wonder if this is one of this things that accidentally works and it should be VUE_APP_API_URL="/api" # yarn run build? because it doesn't really make sense in the way it's used to build URLs. @skullydazed <https://github.com/skullydazed> was this intentional?
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/qmk/qmk_configurator/pull/1015#pullrequestreview-802953669>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAFSNO42FKEOKDSGWOYXHHDULKXUNANCNFSM5HPXUMNQ>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> 

-- 
Hugo Osvaldo Barrera

yanfali commented:
I tried building this on an M1 Mac and it didn't work, but I'm sure it's a node ARM issue. Oh well :/ Seems to have worked just fine for the build itself. I need to try this on an Intel box.

yanfali commented:
@WhyNotHugo OK, I tested it, seems to work just fine. Please just clean up that file removal and I will merge. Thank you!

WhyNotHugo commented:
> Please just clean up that file removal and I will merge. Thank you!

The what?

WhyNotHugo reviewed (commented):
Oh, didn't notice these comments were still draft.

skullydazed commented on Dockerfile:
```diff
@@ -1,31 +1,18 @@
 # Build configurator
-RUN /bin/bash -i /qmk_configurator/bin/docker_build.sh
+RUN yarn install
+ENV VUE_APP_API_URL=/api
+RUN yarn run build
```
There is still a change in behavior here. What you've done is set the environment variable `VUE_APP_API_URL` for all subsequent commands, including the app itself, not just the build. This may or may not be ok, @yanfali would need to weigh in on that.

WhyNotHugo replied:
This line is the last command for this container build. The following line starts a new container build (this is a multistage build), so this `ENV` does not apply to anything else.

WhyNotHugo commented:
FWIW, this image isn't the best for iterating when developing, since it requires a full rebuild after making any changes. It works for quickly creating and environment and running the current version, so it might make sense to actually split it in two.

I do want to keep this PR scoped tho, so I'll look into that separately.

skullydazed commented:
The current docker build is not intended for development, but for deployment. It's not surprising it's less than optimal.

WhyNotHugo commented:
I assumed it was the case. It was very handy to get it running locally.

WhyNotHugo commented:
Any further further feedback here?

I’d like to follow-up with a development-specific container, mostly since I later want to jump onto making improvements onto configurator itself.

yanfali commented:
> Any further further feedback here?
> 
> I’d like to follow-up with a development-specific container, mostly since I later want to jump onto making improvements onto configurator itself.

Please remove yarn.list as it's now longer used.

WhyNotHugo commented:
Good catch. Dropped.

WhyNotHugo commented:
I've also dropped the custom template code in favour of the logic shipped with nginx's docker image. See db74a64a3de0acd9bbba3b5be787c22b72354728

Not sure if including that in this PR is desirable or not.

yanfali commented:
> I've also dropped the custom template code in favour of the logic shipped with nginx's docker image. See [db74a64](https://github.com/qmk/qmk_configurator/commit/db74a64a3de0acd9bbba3b5be787c22b72354728)
> 
> Not sure if including that in this PR is desirable or not.

Skully put that in there, and there's comments about caching. I'm not sure what you've done us equivalent. There's some 404 handling in there too. @skullydazed

WhyNotHugo commented:
404 handling is still there, but let's leave that change out-of-scope so as not to delay this PR.

yanfali commented:
This looks largely the same as the version I tested previously. @WhyNotHugo any objection if we change from ubuntu:latest to debian:11 ? Thanks. I'm not a fan of ubuntu and they tend to be worse for services.

WhyNotHugo commented:
I don't think you can use Debian on Github Actions. Microsoft mostly aligns with Ubuntu, in general, and in this case:

https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idruns-on

yanfali commented:
> I don't think you can use Debian on Github Actions. Microsoft mostly aligns with Ubuntu, in general, and in this case:
> 
> https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobsjob_idruns-on

Oh that's a shame.

yanfali commented:
Thank you for your contribution
